### ISSUE TITLE
[bug] Tauri 2.1.1 & 1.8.1 is affected by glib-rs 0.15 security vulnerability

### ISSUE THREAD

**Issue by ilyagr** [tauri-apps/tauri#12048]
### Describe the bug

The `tauri` 1.8.1 Rust package currently requires glib@0.15 on Linux, but versions of glib (the Rust bindings) >=0.15 and <0.20 are affected by https://github.com/advisories/GHSA-wrw7-89jp-8q8g . I believe (but am not sure) that the nightly version of Tauri 1 probably depends on glib@0.18, which is also bad.

GitHub informed me about this in https://github.com/ilyagr/diffedit3/security/dependabot/10.

### Expected behavior

It'd be great if there was a tauri 1.18.2 that could work with glib 0.20 :)

### Full `tauri info` output

```text
N/A
```

**Comment by FabianLars** [comment]
Since the gtk3 bindings are unmaintained I think this is a `wontfix` sadly. (We don't use glib directly ourselves)

**Comment by ilyagr** [comment]
If the gtk3 bindings are never made compatible with glib 0.20, another option might be to downgrade to gtk 0.14.3. Unless, of course, another security bug appears that affects older versions of some gtk dependency...

**Comment by ilyagr** [comment]
Actually, Tauri 2.1.1 also seems affected. It seems to be using `gtk v0.18.2` which seems to depend on  `glib v0.18`. I'm surprised it's not showing up in the Security tab of this repo; I guess you didn't enable dependency-based security warning.

Cc: https://github.com/tauri-apps/tauri/issues/7335

**Comment by FabianLars** [comment]
We do see them in the Security tab, but only those with repo access can see that (afaik github doesn't allow us to make this public?).
And I think there's no gh issue because this is still not resolved: https://github.com/rustsec/audit-check/issues/8

I doubt anyone in the community would be interested in forking the gtk3 bindings so i'll close this in favor of #7335 (but only because this unsound issue doesn't seem to affect us)

**Comment by ilyagr** [comment]
Thanks for looking into this!

I didn't understand what you said about https://github.com/rustsec/audit-check/issues/8 and how the check should be behaving, nor why this "wouldn't affect us", so I'd appreciate more details if you are feeling like explaining it. But you are the expert here, so I trust your conclusion that Tauri for Linux not on fire. Glad to hear it!

**Comment by FabianLars** [comment]
> I didn't understand what you said about https://github.com/rustsec/audit-check/issues/8 and how the check should be behaving

We have an audit workflow that opens public issues like this one https://github.com/tauri-apps/tauri/issues/11641 but it's not working for `unsound` issues due to a bug in the linked github action.

> nor why this "wouldn't affect us"

This was about the unsound issue itself. We've been using the affected glib versions for 3? years (since the day 0.15 was released basically) without any issues. This is good enough in my eyes to not have us spend resources that we really do not have on forking&maintaining the gtk3 crates.

**Comment by velocitysystems** [comment]
# Analysis: glib-rs Vulnerability

## Problem

The issue reports that Tauri depends on **glib 0.18.5** (via gtk 0.18 and webkit2gtk 2.0.1), which is affected by **[GHSA-wrw7-89jp-8q8g](https://github.com/advisories/GHSA-wrw7-89jp-8q8g)** - an unsoundness bug in `VariantStrIter` that can cause NULL pointer dereferences.

**Affected versions**: glib >= 0.15 and < 0.20

### Vulnerability Details

The `VariantStrIter::impl_get` function was unsound - an immutable reference `&p` to a `*mut libc::c_char` pointer was passed to a C function that mutates the pointer in-place. After recent Rust compiler optimizations, these writes are disregarded, causing NULL pointer dereferences when `std::ffi::CStr::from_ptr` is called.

## Current State

From `Cargo.lock`:
- **glib**: 0.18.5 ❌ (vulnerable)
- **gtk**: 0.18.2
- **webkit2gtk**: 2.0.1

## Challenges

The dependency chain is:
```
tauri-runtime-wry → gtk 0.18 + webkit2gtk 2.0 → glib 0.18
```

**gtk3-rs (gtk 0.18) is unmaintained upstream** and will not receive updates to use glib 0.20. The gtk-rs project has moved to GTK4 bindings only.

## What Can Be Done

### Option 1: Upstream Fix (Ideal but Blocked)

Upgrade to glib 0.20+ requires:
- gtk-rs to release gtk 0.19+ with glib 0.20 support — **not happening** (gtk3-rs is EOL)
- webkit2gtk-rs (Tauri's fork) to update to glib 0.20 — requires gtk 0.19+

### Option 2: Migrate to GTK4 (Long-term)

This would be a major undertaking requiring:
- Porting webkit2gtk-rs to use gtk4-rs
- Significant changes to tauri-runtime-wry
- Work completed in #14684 but requires full code-review, testing and compliance review

### Option 3: Assess Actual Risk (Pragmatic)

The vulnerability is in `glib::VariantStrIter`. If Tauri doesn't use this specific iterator, the practical risk may be low. Worth auditing whether Tauri's code paths actually trigger the vulnerable code.

### Option 4: Workaround via Cargo Patch

Users could potentially patch glib in their own projects, but this is fragile and not a real fix.

### Option 5: Fork and Backport Fix

Fork gtk-rs 0.18 and backport the glib fix from 0.20. This is maintenance-heavy but may be the only viable short-term solution.

## Recommendation

1. **Short-term**: Audit if `VariantStrIter` is actually used in Tauri's Linux code paths
2. **Medium-term**: Consider if webkit2gtk-rs can be updated independently to use newer glib (may require forking gtk-rs 0.18 with the fix backported)
3. **Long-term**: Plan migration path to GTK4 for Tauri v3

## References

- [GitHub Issue #12048](https://github.com/tauri-apps/tauri/issues/12048)
- [GHSA-wrw7-89jp-8q8g](https://github.com/advisories/GHSA-wrw7-89jp-8q8g)
- [Fedora: Retire gtk3-rs](https://fedoraproject.org/wiki/Changes/Retire_gtk3-rs,_gtk-rs-core_v0.18,_and_gtk4-rs_v0.7)
